milabench/validation/error.py

```diff
@@ -0,0 +1,103 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import List
+
+from .validation import ValidationLayer, BenchLogEntry
+
+
+@dataclass
+class PackError:
+    """Error messages received from a running pack"""
+
+    stderr: List[str] = field(default_factory=list)
+    code: int = 0
+    message: str = None
+    early_stop: bool = False
+
+
+def _extract_traceback(lines):
+    output = []
+    traceback = False
+
+    for line in lines:
+        line = line.rstrip()
+
+        if "During handling of the above exception" in line:
+            # The exceptions that happened afterwards are not relevant
+            break
+
+        if "Traceback" in line:
+            traceback = True
+
+        if traceback and line != "":
+            output.append(line)
+
+    return output
+
+
+class _Layer(ValidationLayer):
+    """Capture all error event and save them to generate a summary"""
+
+    def __init__(self, **kwargs) -> None:
+        self.errors = defaultdict(PackError)
```

Comment:
The layers almost all have some tag-indexed defaultdict, I feel like there's something about this pattern that could be abstracted, maybe have a method to initialize for a new tag, and then have on_event take that value for the current tag and return the next one.